### PR TITLE
Report Kotlin build script compile avoidance issues as warnings

### PR DESCRIPTION
**Pull request by rieske**
Follow-up on https://github.com/gradle/gradle/pull/14997

When ABI extraction for compile avoidance fails, we revert to using the full hash of the jar in question. This effectively disables compile avoidance for that classpath entry/jar. With https://github.com/gradle/gradle/pull/14997 this is a silent fallback, only a debug message will be logged, something that a regular user won't spot, which is not helpful in addressing the issue.

This change adds a hook that allows to inject reporting logic for classpath hash computation issues and injects such reporter for Kotlin compile classpath fingerprinter. This reporter currently naively tries to filter out external libraries (because users usually have little control on their content and they are less likely to change, so them not participating in compile avoidance is OK-ish) and warn users about compile avoidance issues in local build logic code.

### CODE REVIEW

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinCompileClasspathFingerprinter.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    override fun report(zipFileSnapshot: RegularFileSnapshot, e: Exception) {
+        val jarName = zipFileSnapshot.name
+        if (e is CompileAvoidanceException && !isExternalDependency(jarName)) {
+            logger.warn("Can not use Kotlin build script compile avoidance with {}: {}", jarName, e.message)
+        }
+    }
+
+    private
+    fun isExternalDependency(jarFileName: String) =
+        jarFileName.matches(".*\\d\\.jar".toRegex())
```
Comment (rieske) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinCompileClasspathFingerprinter.kt:66]: This is not a good and correct way to detect if a dependency is an external one as opposed to buildSrc/included build jar.

Comment (rieske) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinCompileClasspathFingerprinter.kt:66]: If we wanted to only filter out the Gradle related jars, it is possible to filter out Gradle internal dependencies using https://github.com/gradle/gradle/blob/b34e91753ed81aa16360b700e179741da005c780/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassPathProvider.kt#L107-L109 here, however, this does not capture the ones from the `kotlin-dsl` plugin. And I'd say we want to exclude not only Gradle+Kotlin DSL jars from warning that they won't participate in compile avoidance, but rather do it for every jar that comes as an external dependency.

Perhaps the existing naive approach is good enough.

Open to suggestions.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinCompileClasspathFingerprinter.kt [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    }
+}
+
+
+private
+class CompileAvoidanceExceptionReporter : ZipHasher.HashingExceptionReporter {
+    override fun report(zipFileSnapshot: RegularFileSnapshot, e: Exception) {
+        val jarName = zipFileSnapshot.name
+        if (e is CompileAvoidanceException && !isExternalDependency(jarName)) {
+            if (isExternalDependency(jarName)) {
```
Comment (big-guy) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinCompileClasspathFingerprinter.kt:60]: Isn't this always false? (we check isExternalDependency above)

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinCompileClasspathFingerprinter.kt (hunk 1 above)
Comment (big-guy) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinCompileClasspathFingerprinter.kt:70]: I think the easier thing to do for now is to not try to warn the user, make this an info level log and see if there are complaints. The real fix is to make inline methods work properly or narrow their impact more.

So we can just not worry about detecting this at all and stick with just some logging.

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinCompileClasspathFingerprinter.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                logger.info("Cannot use Kotlin build script compile avoidance with {}: {}", jarName, e.message)
```
Comment (big-guy) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinCompileClasspathFingerprinter.kt:61]: maybe we should log the full path?

**Review by big-guy — CHANGES_REQUESTED** [review]

**Pull request merged**